### PR TITLE
Improve warnings: use color and sane defaults

### PR DESCRIPTION
**Pull request by imLew**
**Proposed changes**:
- Improve warnings: use color and sane defaults when emitting warnings for our users
- Makes sure the warning messages are colored and use a sane stack level for the warning message (e.g. by default this would be the line emitting the warning, which in most cases doesn't make sense).
- If the code line is not relevant (e.g. if the warning is about an intent that should be listed in the domain but is not) we add a link to some relevant documentation page.
- related to https://github.com/RasaHQ/rasa/issues/4356

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by wochinge** [comment]
@tmbo Who is taking this over?

**Comment by tmbo** [comment]
Can you please take that over as a squad? 

I think it is unrealistic to set the warnings filter to `error` just yet (due to the strange errors we get with that), but here is what I think we should keep:
- helper that sets default values for category and stacklevel
- optional: red formatting of these messages in the log if possible

**Comment by wochinge** [comment]
You mean as production squad? ok

**Comment by wochinge** [comment]
@alwx I think @tmbo is currently also doing something here. Best keep off for now

**Comment by tmbo** [comment]
should have assigned me, sorry for the confusion

File: rasa/core/tracker_store.py [hunk 1]
```diff
@@ -722,9 +713,6 @@ def keys(self) -> Iterable[Text]:
     def retrieve(self, sender_id: Text) -> Optional[DialogueStateTracker]:
         """Create a tracker from all previously stored events."""
 
-        import sqlalchemy as sa
```
Comment (tmbo) [inline rasa/core/tracker_store.py]: check if this import is needed

File: rasa/nlu/classifiers/keyword_intent_classifier.py [hunk 2]
```diff
@@ -152,8 +155,14 @@ def load(
...
-                    "{keyword_file} does not exist."
+                raise_warning(
+                    f"Failed to load key word file for `IntentKeywordClassifier`, "
+                    f"maybe {keyword_file} does not exist?",
                 )
-        return cls(meta, intent_keyword_map)
+                intent_keyword_map = None
+            return cls(meta, intent_keyword_map)
+        else:
+            raise Exception(
```
Comment (tmbo) [inline rasa/nlu/classifiers/keyword_intent_classifier.py]: converted this to an exception to stay in line with other components

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 3]
```diff
@@ -108,7 +110,12 @@ def train(
 
             self.clf = self._create_classifier(num_threads, y)
 
-            self.clf.fit(X, y)
+            with warnings.catch_warnings():
```
Comment (tmbo) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py]: ignore sklearn warnings

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 4]
```diff
@@ -140,6 +147,7 @@ def _create_classifier(
             cv=cv_splits,
             scoring=self.component_config["scoring_function"],
             verbose=1,
+            iid=False,
```
Comment (tmbo) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py]: prevents deprecation warning

File: rasa/utils/common.py [hunk 5]
```diff
@@ -295,3 +299,47 @@ def _lazyprop(self):
         return getattr(self, attr_name)
 
     return _lazyprop
+
+
+def raise_warning(
```
Comment (tmbo) [inline rasa/utils/common.py]: implementation for raise_warning

File: setup.cfg [hunk 6]
```diff
@@ -13,6 +13,7 @@ codestyle_exclude =
     rasa/core/policies/__init__.py
 filterwarnings =
     ignore::ResourceWarning:ruamel[.*]
+    #error
```
Comment (tmbo) [inline setup.cfg]: this should be activated later on to fail on warnings, doesn't work yet though

File: tests/cli/test_utils.py [hunk 7]
```diff
@@ -98,31 +97,9 @@ def test_validate_with_invalid_directory_if_default_is_valid(caplog: LogCaptureF
...
-    assert "does not exist" in record[0].message.args[0]
+    assert "does not seem to exist" in record[0].message.args[0]
 
 
 def test_print_error_and_exit():
     with pytest.raises(SystemExit):
         rasa.cli.utils.print_error_and_exit("")
-
-
-def test_logging_capture(caplog: LogCaptureFixture):
```
Comment (tmbo) [inline tests/cli/test_utils.py]: tests removed since that function (`assert_log_emitted `) isn't used anywhere

**Comment by tmbo** [comment]
This is ready for review and should go into the 1.7 release (cc @federicotdn).

@degiz can you please give this a review? it looks like a lot of changes but most of them are import reorg (thought about adding isort to prevent this from happening, but different story). I've commented on the lines where I did more than just rename `warnings.warn`.

**Comment by tmbo** [comment]
(@degiz ideally today or early tomorrow so that we can get it into the release - let me know if you have to much on your plate)

File: data/test_domains/default_with_slots_and_no_actions.yml [hunk 8]
```diff
@@ -18,7 +18,7 @@ slots:
   name:
     type: text
 
-templates:
+responses:
```
Comment (tmbo) [inline data/test_domains/default_with_slots_and_no_actions.yml]: leftover from an earlier rename (created a warning that it is deprecated, that is why I changed it)

File: rasa/core/agent.py [hunk 9]
```diff
@@ -593,8 +593,8 @@ def toggle_memoization(self, activate: bool) -> None:
     def continue_training(
         self, trackers: List[DialogueStateTracker], **kwargs: Any
     ) -> None:
-        warnings.warn(
-            "Continue training will be removed in the next release. It won't be "
+        raise_warning(
+            "Continue training will be removed in the 2.0 release. It won't be "
```
Comment (degiz) [inline rasa/core/agent.py]: optional: maybe it makes sense to mark the function as deprecated? is it exposed to the users?

File: rasa/core/agent.py [hunk 10]
```diff
@@ -798,9 +798,9 @@ def persist(self, model_path: Text, dump_flattened_stories: bool = False) -> Non
         """Persists this agent into a directory for later loading and usage."""
 
         if dump_flattened_stories:
-            warnings.warn(
+            raise_warning(
                 "The `dump_flattened_stories` argument will be removed from "
-                "`Agent.persist` in the next release. Please dump your "
+                "`Agent.persist` in the 2.0 release. Please dump your "
```
Comment (degiz) [inline rasa/core/agent.py]: optional: same here.

File: rasa/core/brokers/file_producer.py [hunk 11]
```diff
@@ -1,13 +1,14 @@
...
 class FileProducer(FileEventBroker):
-    warnings.warn(
+    raise_warning(
         "The `FileProducer` class is deprecated, please inherit from "
         "`FileEventBroker` instead. `FileProducer` will be removed in "
         "future Rasa versions.",
-        DeprecationWarning,
-        stacklevel=2,
+        FutureWarning,
+        docs="/api/event-brokers/",
```
Comment (degiz) [inline rasa/core/brokers/file_producer.py]: optional: we could have this API sections as constants somewhere just not to update all of them every time.

File: rasa/core/training/dsl.py [hunk 12]
```diff
@@ -412,10 +415,12 @@ def add_checkpoint(self, name: Text, conditions: Optional[Dict[Text, Any]]) -> N
...
         intent_name = utterance.intent.get("name")
         if intent_name not in self.domain.intents:
-            warnings.warn(
+            raise_warning(
                 f"Found unknown intent '{intent_name}' on line {line_num}. "
                 "Please, make sure that all intents are "
-                "listed in your domain yaml."
+                "listed in your domain yaml.",
+                UserWarning,
+                docs="https://rasa.com/docs/rasa/core/domains/",
```
Comment (degiz) [inline rasa/core/training/dsl.py]: so here we give full URL to the doc compared to relative one before?
What's the desired way?

File: rasa/core/validator.py [hunk 13]
```diff
@@ -141,8 +142,12 @@ def verify_utterances(self, ignore_warnings: bool = True) -> bool:
         for action in actions:
             if action.startswith(UTTER_PREFIX):
                 if action not in utterance_templates:
-                    warnings.warn(
-                        f"There is no template for utterance '{action}'.", stacklevel=2
+                    raise_warning(
+                        f"There is no template for the utterance action '{action}'."
```
Comment (degiz) [inline rasa/core/validator.py]:
```suggestion
                        f"There is no template for the utterance action '{action}'. "
```

File: rasa/nlu/classifiers/keyword_intent_classifier.py [hunk 14]
```diff
@@ -86,13 +88,14 @@ def _validate_keyword_map(self) -> None:
                     and intent1 != intent2
                 ):
                     ambiguous_mappings.append((intent1, keyword1))
-                    warnings.warn(
+                    raise_warning(
                         f"Keyword '{keyword1}' is a keyword of intent '{intent1}', "
                         f"but also a substring of '{keyword2}', which is a "
                         f"keyword of intent '{intent2}."
                         f" '{keyword1}' will be removed from the list of keywords.\n"
                         "Remove (one of) the conflicting keywords from the"
```
Comment (degiz) [inline rasa/nlu/classifiers/keyword_intent_classifier.py]:
```suggestion
                        f"Remove (one of) the conflicting keywords from the"
```

File: rasa/nlu/classifiers/keyword_intent_classifier.py [hunk 15]
```diff
@@ -86,13 +88,14 @@ def _validate_keyword_map(self) -> None:
...
-                        " training data."
+                        " training data.",
```
Comment (degiz) [inline rasa/nlu/classifiers/keyword_intent_classifier.py]:
```suggestion
                        f" training data.",
```

File: rasa/nlu/classifiers/keyword_intent_classifier.py (hunk 14 above)
Comment (degiz) [reply rasa/nlu/classifiers/keyword_intent_classifier.py]: looks like we're adding "f" for all sub-lines of multi strings, so we need it here as well.

File: rasa/nlu/featurizers/dense_featurizer/convert_featurizer.py [hunk 16]
```diff
@@ -174,11 +176,12 @@ def train(
     ) -> None:
 
         if config is not None and config.language != "en":
-            warnings.warn(
+            raise_warning(
                 f"Since ``ConveRT`` model is trained only on an english "
                 f"corpus of conversations, this featurizer should only be "
                 f"used if your training data is in english language. "
-                f"However, you are training in '{config.language}'."
+                f"However, you are training in '{config.language}'.",
```
Comment (degiz) [inline rasa/nlu/featurizers/dense_featurizer/convert_featurizer.py]:
```suggestion
                f"However, you are training in '{config.language}'. ",
```

File: rasa/utils/common.py [hunk 17]
```diff
@@ -295,3 +299,47 @@ def _lazyprop(self):
...
+
+    def should_show_source_line():
+        if "stacklevel" not in kwargs:
+            if category == UserWarning or category is None:
+                return False
+            if category == FutureWarning:
+                return False
+        return True
+
+    def formatwarning(message, category, filename, lineno, line=None):
```
Comment (degiz) [inline rasa/utils/common.py]: can we add types here?

File: tests/core/test_processor.py [hunk 18]
```diff
@@ -78,13 +78,14 @@
     with pytest.warns(UserWarning) as record:
         default_processor._log_unseen_features(parsed)
     assert len(record) == 2
+
     assert (
-        record[0].message.args[0]
-        == "Interpreter parsed an intent 'dislike' that is not defined in the domain."
+        record[0].message.args[0].startswith("Interpreter parsed an intent 'dislike'")
```
Comment (degiz) [inline tests/core/test_processor.py]: Thinking out loud: That can be tough in the future to have the warring text in different places: in the original business logic file and in the test.

File: rasa/core/agent.py (hunk 9 above)
Comment (tmbo) [reply rasa/core/agent.py]: mhm not sure what you mean, there isn't a `@Deprecated` like in java 😅

**Comment by ricwo** [comment]
pytest has introduced coloured test summaries with `5.3.0`: https://docs.pytest.org/en/latest/changelog.html#id29

we could upgrade to latest v 5.3.4

File: rasa/core/training/dsl.py (hunk 12 above)
Comment (tmbo) [reply rasa/core/training/dsl.py]: fixed this, it is full URL now that we have constants (before it was relative, which is great - now we can also link to idk, some tensorflow issue or something)

**Comment by tmbo** [comment]
@ricwo that sounds great we should def do that, but would do that after the release :+1:

File: tests/core/test_processor.py (hunk 18 above)
Comment (tmbo) [reply tests/core/test_processor.py]: yes very true, I think this test shouldn't check for the actual message, checking if a warning is raised should be enough :+1:

File: rasa/utils/common.py (hunk 17 above)
Comment (tmbo) [reply rasa/utils/common.py]: yes :+1:

**Pull request merged**